File: 3.x.x/crates-utopia/span/src/exporter/sentry.rs
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::{json, Map, Value};
use thiserror::Error;

use crate::attr::AttrValue;
use crate::error::SpanError;
use crate::exporter::{Exporter, SentryField, SentryLevel};
use crate::level::Level;
use crate::php_url::parse_url;
use crate::span::Span;

const EXPORT_LEVELS: [Level; 3] = [Level::Warn, Level::Error, Level::Fatal];
const MAX_CHAIN_DEPTH: usize = 10;
const HANDLED_HTTP_KEYS: [&str; 4] = [
    "http.url",
    "http.method",
    "http.query",
    "http.response.status_code",
];

type Sampler = Box<dyn Fn(&Span) -> bool + Send + Sync>;
type Classifier = Box<dyn Fn(&str) -> SentryField + Send + Sync>;

/// PHP `InvalidArgumentException` from the Sentry exporter constructor.
#[derive(Debug, Error, Clone, PartialEq, Eq)]
pub enum SentryError {
    #[error("Sentry DSN is required")]
    DsnRequired,
    #[error("Invalid Sentry DSN")]
    InvalidDsn,
    #[error("Invalid Sentry DSN: must include public key, host, and project ID")]
    IncompleteDsn,
}

/// Exports warning-or-higher spans to Sentry Issues (PHP `Exporter\Sentry`).
pub struct Sentry {
    dsn: String,
    endpoint: String,
    public_key: String,
    environment: Option<String>,
    release: Option<String>,
    server_name: Option<String>,
    classifier: Classifier,
    sampler: Sampler,
}

impl std::fmt::Debug for Sentry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Sentry")
            .field("endpoint", &self.endpoint)
            .field("environment", &self.environment)
            .finish_non_exhaustive()
    }
}

impl Sentry {
    pub fn new(dsn: impl Into<String>) -> Result<Self, SentryError> {
        Self::new_with(None, dsn, None, None, None, None)
    }

    pub fn new_with(
        sampler: Option<Sampler>,
        dsn: impl Into<String>,
        environment: Option<String>,
        release: Option<String>,
        server_name: Option<String>,
        classifier: Option<Classifier>,
    ) -> Result<Self, SentryError> {
        let dsn = dsn.into();
        let user_sampler = sampler;
        let composed: Sampler = Box::new(move |span: &Span| {
            let level = span
                .get("level")
                .and_then(|v| v.as_str().map(str::to_string))
                .and_then(|s| Level::try_from_attr(&s));
            let Some(level) = level else {
                return false;
            };
            if !EXPORT_LEVELS.contains(&level) {
                return false;
            }
            match user_sampler.as_ref() {
                Some(sampler) => sampler(span),
                None => true,
            }
        });
        if dsn.is_empty() {
            return Err(SentryError::DsnRequired);
        }
        let parsed = parse_url(&dsn).ok_or(SentryError::InvalidDsn)?;
        let public_key = parsed.user.unwrap_or_default();
        let host = parsed.host.unwrap_or_default();
        let project_id = parsed
            .path
            .unwrap_or_default()
            .trim_start_matches('/')
            .to_string();
        if public_key.is_empty() || host.is_empty() || project_id.is_empty() {
            return Err(SentryError::IncompleteDsn);
        }
        let scheme = parsed.scheme.unwrap_or_else(|| "https".to_string());
        let port = parsed.port.map(|p| format!(":{p}")).unwrap_or_default();
        let endpoint = format!("{scheme}://{host}{port}/api/{project_id}/envelope/");
        Ok(Self {
            dsn,
            endpoint,
            public_key,
            environment,
            release,
            server_name,
            classifier: classifier.unwrap_or_else(|| Box::new(|_| SentryField::Context)),
            sampler: composed,
        })
    }

    pub fn build_envelope(&self, span: &Span) -> Option<String> {
        let error = span.get_error()?;
        let attributes = span.get_attributes();
        let attr = |key: &str| {
            attributes
                .iter()
                .find(|(k, _)| k == key)
                .map(|(_, v)| v.clone())
        };

        let trace_id = attr("span.trace_id")
            .and_then(|v| v.as_str().map(str::to_string))
            .unwrap_or_default();
        let span_id = attr("span.id")
            .and_then(|v| v.as_str().map(str::to_string))
            .unwrap_or_default();
        let parent_id = attr("span.parent_id").and_then(|v| v.as_str().map(str::to_string));
        let started_at = attr("span.started_at")
            .and_then(|v| v.as_f64())
            .unwrap_or_else(unix_seconds);
        let finished_at = attr("span.finished_at")
            .and_then(|v| v.as_f64())
            .unwrap_or_else(unix_seconds);
        let action = span.get_action();

        let mut trace_context = json!({
            "trace_id": trace_id,
            "span_id": span_id,
        });
        if let Some(parent) = parent_id {
            trace_context["parent_span_id"] = Value::String(parent);
        }

        let header = json!({
            "event_id": trace_id.replace('-', ""),
            "sent_at": date_c(),
            "dsn": self.dsn,
        });
        let item_header = json!({
            "type": "event",
            "content_type": "application/json",
        });

        let mut contexts = Map::new();
        contexts.insert("trace".into(), trace_context);
        contexts.insert(
            "runtime".into(),
            json!({
                "name": "rust",
                "version": option_env!("RUSTC_SEMVER").unwrap_or("unknown"),
            }),
        );

        let request = build_request(&attributes);
        let response = build_response(&attributes);
        if !response.is_empty() {
            contexts.insert("response".into(), Value::Object(response));
        }

        let (tags, custom, extra) = self.classify(&attributes);
        if !custom.is_empty() {
            contexts.insert("custom".into(), Value::Object(custom));
        }

        let level = attr("level")
            .and_then(|v| v.as_str().map(str::to_string))
            .and_then(|s| Level::try_from_attr(&s))
            .unwrap_or(Level::Error);
        let handled_override = match attr("span.handled") {
            Some(AttrValue::Bool(value)) => Some(value),
            _ => None,
        };

        let mut payload = json!({
            "level": SentryLevel::from_span(level).as_str(),
            "platform": "rust",
            "sdk": {
                "name": "utopia-span",
                "version": env!("CARGO_PKG_VERSION"),
            },
            "start_timestamp": started_at,
            "timestamp": finished_at,
            "transaction": action,
            "message": error.message,
            "contexts": contexts,
            "exception": {
                "values": build_exception_values(&error, level, handled_override),
            },
        });

        if !tags.is_empty() {
            payload["tags"] = Value::Object(tags);
        }
        if !extra.is_empty() {
            payload["extra"] = Value::Object(extra);
        }
        if !request.is_empty() {
            payload["request"] = Value::Object(request);
        }
        if let Some(environment) = &self.environment {
            payload["environment"] = Value::String(environment.clone());
        }
        if let Some(release) = &self.release {
            payload["release"] = Value::String(release.clone());
        }
        if let Some(server_name) = &self.server_name {
            payload["server_name"] = Value::String(server_name.clone());
        }

        let header = serde_json::to_string(&header).ok()?;
        let item_header = serde_json::to_string(&item_header).ok()?;
        let payload = serde_json::to_string(&payload).ok()?;
        Some(format!("{header}\n{item_header}\n{payload}"))
    }

    fn classify(
        &self,
        attributes: &[(String, AttrValue)],
    ) -> (Map<String, Value>, Map<String, Value>, Map<String, Value>) {
        let mut tags = Map::new();
        let mut contexts = Map::new();
        let mut extra = Map::new();
        for (key, value) in attributes {
            if key.starts_with("span.")
                || key == "level"
                || HANDLED_HTTP_KEYS.contains(&key.as_str())
            {
                continue;
            }
            match (self.classifier)(key) {
                SentryField::Tag => {
                    let rendered = value.display();
                    let truncated: String = rendered.chars().take(200).collect();
                    tags.insert(key.clone(), Value::String(truncated));
                }
                SentryField::Context => {
                    contexts.insert(key.clone(), value.to_json());
                }
                SentryField::Extra => {
                    extra.insert(key.clone(), value.to_json());
                }
            }
        }
        (tags, contexts, extra)
    }
}

impl Exporter for Sentry {
    fn sample(&self, span: &Span) -> bool {
        (self.sampler)(span)
    }

    fn export(&self, span: &Span) {
        let Some(envelope) = self.build_envelope(span) else {
            return;
        };
        let client = match reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_millis(1000))
            .connect_timeout(std::time::Duration::from_millis(500))
            .build()
        {
            Ok(client) => client,
            Err(err) => {
                eprintln!("Sentry exporter: Failed to initialize curl: {err}");
                return;
            }
        };
        let auth = format!("Sentry sentry_version=7, sentry_key={}", self.public_key);
        match client
            .post(&self.endpoint)
            .header("Content-Type", "application/x-sentry-envelope")
            .header("X-Sentry-Auth", auth)
            .body(envelope)
            .send()
        {
            Ok(response) => {
                let status = response.status().as_u16();
                if status >= 400 {
                    let body = response.text().unwrap_or_default();
                    eprintln!("Sentry exporter: HTTP {status} - {body}");
                }
            }
            Err(err) => eprintln!("Sentry exporter: {err}"),
        }
    }
}

fn build_exception_values(
    error: &SpanError,
    level: Level,
    handled_override: Option<bool>,
) -> Vec<Value> {
    let mut chain = Vec::new();
    let mut current = Some(error);
    while let Some(err) = current {
        if chain.len() >= MAX_CHAIN_DEPTH {
            break;
        }
        chain.push(err);
        current = err.previous.as_deref();
    }
    let handled = handled_override.unwrap_or(level != Level::Fatal);
    let chained = chain.len() > 1;
    let mut values = Vec::new();
    for (id, exception) in chain.iter().enumerate() {
        let mut mechanism = json!({
            "type": "generic",
            "handled": handled,
        });
        if chained {
            mechanism["exception_id"] = json!(id);
            if id > 0 {
                mechanism["parent_id"] = json!(id - 1);
                mechanism["source"] = json!("__previous__");
            }
        }
        let mut value = json!({
            "type": exception.type_name,
            "value": exception.message,
            "stacktrace": { "frames": build_frames(exception) },
            "mechanism": mechanism,
        });
        if let Some(module) = exception.module() {
            value["module"] = Value::String(module.to_string());
        }
        values.push(value);
    }
    values.reverse();
    values
}

fn build_frames(error: &SpanError) -> Vec<Value> {
    let mut frames = Vec::new();
    for frame in error.frames.iter().rev() {
        let Some(file) = &frame.file else {
            continue;
        };
        if file == &error.file && frame.line == Some(error.line) {
            continue;
        }
        let mut obj = json!({
            "filename": file,
            "lineno": frame.line.unwrap_or(0),
            "in_app": in_app(file),
        });
        if !frame.function.is_empty() {
            obj["function"] = Value::String(frame.function.clone());
        }
        frames.push(obj);
    }
    let mut throw_site = json!({
        "filename": error.file,
        "lineno": error.line,
        "in_app": in_app(&error.file),
    });
    if let Some(frame) = error.frames.first() {
        if !frame.function.is_empty() {
            throw_site["function"] = Value::String(frame.function.clone());
        }
    }
    frames.push(throw_site);
    frames
}

fn in_app(file: &str) -> bool {
    !file.contains("/vendor/") && !file.contains("/.cargo/") && !file.contains("/rustc/")
}

fn build_request(attributes: &[(String, AttrValue)]) -> Map<String, Value> {
    let mut request = Map::new();
    if let Some(AttrValue::String(url)) = get_attr(attributes, "http.url") {
        request.insert("url".into(), Value::String(url.clone()));
    }
    if let Some(AttrValue::String(method)) = get_attr(attributes, "http.method") {
        request.insert("method".into(), Value::String(method.clone()));
    }
    if let Some(AttrValue::String(query)) = get_attr(attributes, "http.query") {
        request.insert("query_string".into(), Value::String(query.clone()));
    }
    request
}

fn build_response(attributes: &[(String, AttrValue)]) -> Map<String, Value> {
    let mut response = Map::new();
    if let Some(AttrValue::Int(code)) = get_attr(attributes, "http.response.status_code") {
        response.insert("status_code".into(), json!(code));
    }
    response
}

fn get_attr<'a>(attributes: &'a [(String, AttrValue)], key: &str) -> Option<&'a AttrValue> {
    attributes.iter().find(|(k, _)| k == key).map(|(_, v)| v)
}

fn unix_seconds() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

fn date_c() -> String {
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    let (year, month, day, hour, min, sec) = civil_utc(secs);
    format!("{year:04}-{month:02}-{day:02}T{hour:02}:{min:02}:{sec:02}+00:00")
}

fn civil_utc(unix: i64) -> (i32, u32, u32, u32, u32, u32) {
    let day_secs = 86_400;
    let days = unix.div_euclid(day_secs);
    let rem = unix.rem_euclid(day_secs) as u32;
    let hour = rem / 3600;
    let min = (rem % 3600) / 60;
    let sec = rem % 60;
    let (year, month, day) = civil_from_days(days);
    (year, month, day, hour, min, sec)
}

/// Howard Hinnant civil-from-days (UTC).
fn civil_from_days(z: i64) -> (i32, u32, u32) {
    let z = z + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1_460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    (y as i32, m as u32, d as u32)
}

File: packages/span/src/Span/Exporter/Sentry.php
<?php

namespace Utopia\Span\Exporter;

use Closure;
use Composer\InstalledVersions;
use Psr\Http\Client\ClientInterface;
use Utopia\Client\Adapter\Curl\Client as CurlClient;
use Utopia\Client as HttpClient;
use Utopia\Psr7\Method;
use Utopia\Psr7\Request\Factory as RequestFactory;
use Utopia\Span\Exporter\Sentry\Level as SentryLevel;
use Utopia\Span\Level;
use Utopia\Span\Span;

/**
 * Exports warning-or-higher spans to Sentry Issues.
 *
 * Only spans whose level is warning, error, or fatal are sent, at that level.
 * Lower levels (info, debug) are skipped — use the Stdout exporter for those.
 *
 * ## Exceptions
 *
 * The span's error and its full getPrevious() chain (up to 10 links) are sent
 * as chained exception values, so Sentry shows the root cause alongside the
 * reported exception. Each value carries a mechanism whose handled flag comes
 * from a boolean `span.handled` attribute when set; otherwise warning/error
 * spans are marked handled and fatal spans unhandled.
 *
 * ## HTTP Attribute Conventions
 *
 * The following span attributes are mapped to Sentry's request/response structures:
 *
 * Request attributes:
 * - `http.url` - Full request URL
 * - `http.method` - HTTP method (GET, POST, etc.)
 * - `http.query` - Query string
 *
 * Response attributes:
 * - `http.response.status_code` - HTTP response status code
 *
 * ## Attribute Classification
 *
 * By default, all unhandled attributes go to `context` (recommended by Sentry).
 * Use the `$classifier` callback to control where attributes are placed:
 *
 * ```php
 * $exporter = new Sentry($dsn, classifier: function(string $key): SentryField {
 *     return match(true) {
 *         str_starts_with($key, 'tenant.') => SentryField::Tag,
 *         default => SentryField::Context,
 *     };
 * });
 * ```
 */
class Sentry implements Exporter
{
    private static ?string $sdkVersion = null;

    /**
     * Span levels that are exported. Anything below warning (info, debug) is skipped.
     */
    private const array EXPORT_LEVELS = [Level::Warn, Level::Error, Level::Fatal];

    /**
     * Maximum number of chained exceptions (via getPrevious()) sent per event.
     */
    private const int MAX_CHAIN_DEPTH = 10;

    private readonly string $endpoint;
    private readonly string $publicKey;
    private readonly string $projectId;

    /** @var Closure(string): SentryField */
    private readonly Closure $classifier;

    /** @var Closure(Span): bool */
    private readonly Closure $sampler;

    private readonly ClientInterface $client;

    private readonly RequestFactory $requestFactory;

    /**
     * Create a new Sentry exporter.
     *
     * Sentry only exports spans at warning level or above; a custom sampler is composed (AND)
     * with the built-in level filter, so it can further restrict — but not broaden — what is sent.
     *
     * @param Closure(Span): bool|null $sampler Optional additional filter, composed with the level filter.
     * @param string $dsn Sentry DSN (e.g., https://key@sentry.io/123)
     * @param string|null $environment Optional environment name (e.g., 'production')
     * @param string|null $release Optional release/version identifier (e.g., commit hash)
     * @param string|null $serverName Optional server name/identifier
     * @param Closure(string): SentryField|null $classifier Optional callback to classify attributes
     * @param ClientInterface|null $client Optional PSR-18 transport; defaults to `utopia-php/client` with its cURL adapter
     */
    public function __construct(
        ?Closure $sampler = null,
        private readonly string $dsn = '',
        private readonly ?string $environment = null,
        private readonly ?string $release = null,
        private readonly ?string $serverName = null,
        ?Closure $classifier = null,
        ?ClientInterface $client = null,
    ) {
        $this->classifier = $classifier ?? static fn(string $key): SentryField => SentryField::Context;
        $this->client = $client ?? new HttpClient(
            new CurlClient(options: [
                \CURLOPT_TIMEOUT_MS => 1000,
                \CURLOPT_CONNECTTIMEOUT_MS => 500,
            ]),
        );
        $this->requestFactory = new RequestFactory();
        $this->sampler = static function (Span $span) use ($sampler): bool {
            $level = Level::tryFrom((string) $span->get('level'));
            if ($level === null || !\in_array($level, self::EXPORT_LEVELS, true)) {
                return false;
            }
            return !$sampler instanceof \Closure || $sampler($span);
        };
        if ($dsn === '') {
            throw new \InvalidArgumentException('Sentry DSN is required');
        }

        $parsed = parse_url($dsn);

        if ($parsed === false) {
            throw new \InvalidArgumentException('Invalid Sentry DSN');
        }

        $publicKey = $parsed['user'] ?? '';
        $host = $parsed['host'] ?? '';
        $projectId = ltrim($parsed['path'] ?? '', '/');

        if ($publicKey === '' || $host === '' || $projectId === '') {
            throw new \InvalidArgumentException('Invalid Sentry DSN: must include public key, host, and project ID');
        }

        $this->publicKey = $publicKey;
        $this->projectId = $projectId;

        $scheme = $parsed['scheme'] ?? 'https';
        $port = isset($parsed['port']) ? ':' . $parsed['port'] : '';

        $this->endpoint = "{$scheme}://{$host}{$port}/api/{$this->projectId}/envelope/";
    }

    public function sample(Span $span): bool
    {
        return ($this->sampler)($span);
    }

    public function export(Span $span): void
    {
        $envelope = $this->buildEnvelope($span);

        if ($envelope === null) {
            return;
        }

        $this->sendWithClient($this->client, $envelope);
    }

    private function sendWithClient(ClientInterface $client, string $envelope): void
    {
        $request = $this->requestFactory->body(
            Method::POST,
            $this->endpoint,
            $envelope,
            'application/x-sentry-envelope',
            ['X-Sentry-Auth' => "Sentry sentry_version=7, sentry_key={$this->publicKey}"],
        );

        try {
            $response = $client->sendRequest($request);
            $statusCode = $response->getStatusCode();

            if ($statusCode >= 400) {
                error_log("Sentry exporter: HTTP {$statusCode} - {$response->getBody()}");
            }
        } catch (\Throwable $error) {
            error_log('Sentry exporter: ' . $error->getMessage());
        }
    }

    private function buildEnvelope(Span $span): ?string
    {
        $error = $span->getError();

        if (!$error instanceof \Throwable) {
            return null;
        }

        $attributes = $span->getAttributes();

        $traceId = (string) ($attributes['span.trace_id'] ?? '');
        $spanId = (string) ($attributes['span.id'] ?? '');
        $parentId = $attributes['span.parent_id'] ?? null;
        $startedAt = (float) ($attributes['span.started_at'] ?? microtime(true));
        $finishedAt = (float) ($attributes['span.finished_at'] ?? microtime(true));
        $action = $span->getAction();

        $traceContext = [
            'trace_id' => $traceId,
            'span_id' => $spanId,
        ];

        if (\is_string($parentId)) {
            $traceContext['parent_span_id'] = $parentId;
        }

        $header = json_encode([
            'event_id' => str_replace('-', '', $traceId),
            'sent_at' => date('c'),
            'dsn' => $this->dsn,
        ]);

        $itemHeader = json_encode([
            'type' => 'event',
            'content_type' => 'application/json',
        ]);

        $contexts = [
            'trace' => $traceContext,
            'runtime' => [
                'name' => 'php',
                'version' => PHP_VERSION,
                'sapi' => PHP_SAPI,
            ],
        ];

        $request = $this->buildRequest($attributes);
        $response = $this->buildResponse($attributes);

        if ($response !== []) {
            $contexts['response'] = $response;
        }

        [$tags, $customContexts, $extra] = $this->classifyAttributes($attributes);

        if ($customContexts !== []) {
            $contexts['custom'] = $customContexts;
        }

        $level = Level::tryFrom((string) $attributes['level']) ?? Level::Error;

        $payloadData = [
            'level' => SentryLevel::fromSpan($level)->value,
            'platform' => 'php',
            'sdk' => [
                'name' => 'utopia-php/span',
                'version' => self::$sdkVersion ??= InstalledVersions::getVersion('utopia-php/span') ?? 'unknown',
            ],
            'start_timestamp' => $startedAt,
            'timestamp' => $finishedAt,
            'transaction' => $action,
            'message' => $error->getMessage(),
            'contexts' => $contexts,
            'exception' => [
                'values' => $this->buildExceptionValues($error, $level, \is_bool($attributes['span.handled'] ?? null) ? $attributes['span.handled'] : null),
            ],
        ];

        if ($tags !== []) {
            $payloadData['tags'] = $tags;
        }

        if ($extra !== []) {
            $payloadData['extra'] = $extra;
        }

        if ($request !== []) {
            $payloadData['request'] = $request;
        }

        if ($this->environment !== null) {
            $payloadData['environment'] = $this->environment;
        }

        if ($this->release !== null) {
            $payloadData['release'] = $this->release;
        }

        if ($this->serverName !== null) {
            $payloadData['server_name'] = $this->serverName;
        }

        $payload = json_encode($payloadData);

        if ($header === false || $itemHeader === false || $payload === false) {
            return null;
        }

        return "{$header}\n{$itemHeader}\n{$payload}";
    }

    /**
     * Build the Sentry exception values list, following the getPrevious() chain.
     *
     * Values are ordered oldest cause first (root cause at index 0, reported
     * exception last), as required by the Sentry exception interface. Each value
     * carries a mechanism; chained values are linked via exception_id/parent_id
     * so Sentry renders the cause tree.
     *
     * @return array<int, array<string, mixed>>
     */
    private function buildExceptionValues(\Throwable $error, Level $level, ?bool $handledOverride = null): array
    {
        $chain = [];
        for ($current = $error; $current instanceof \Throwable && \count($chain) < self::MAX_CHAIN_DEPTH; $current = $current->getPrevious()) {
            $chain[] = $current;
        }

        // A 'span.handled' attribute states the handling state explicitly; the
        // level is only a fallback heuristic (fatal implies a process-level
        // handler, anything reported at warning/error implies user code).
        $handled = $handledOverride ?? ($level !== Level::Fatal);
        $chained = \count($chain) > 1;
        $values = [];

        // $chain[0] is the reported (outermost) exception — the root of the
        // mechanism tree, so it gets exception_id 0.
        foreach ($chain as $id => $exception) {
            $mechanism = [
                'type' => 'generic',
                'handled' => $handled,
            ];

            if ($chained) {
                $mechanism['exception_id'] = $id;
                if ($id > 0) {
                    $mechanism['parent_id'] = $id - 1;
                    $mechanism['source'] = '__previous__';
                }
            }

            $value = [
                'type' => $exception::class,
                'value' => $exception->getMessage(),
                'stacktrace' => ['frames' => $this->buildFrames($exception)],
                'mechanism' => $mechanism,
            ];

            $namespaceEnd = strrpos($exception::class, '\\');
            if ($namespaceEnd !== false) {
                $value['module'] = substr($exception::class, 0, $namespaceEnd);
            }

            $values[] = $value;
        }

        return array_reverse($values);
    }

    /**
     * Build Sentry stacktrace frames for one throwable, oldest call first,
     * ending with the throw site.
     *
     * @return array<int, array<string, mixed>>
     */
    private function buildFrames(\Throwable $error): array
    {
        $trace = $error->getTrace();

        $frames = [];
        foreach (array_reverse($trace) as $frame) {
            if (!isset($frame['file'])) {
                continue;
            }
            $frames[] = [
                'filename' => $frame['file'],
                'lineno' => $frame['line'] ?? 0,
                'in_app' => !str_contains($frame['file'], '/vendor/'),
                'function' => $this->formatFunction($frame),
            ];
        }

        // PHP traces describe call sites, not the throw site itself — append it.
        // Its enclosing function is the callee recorded in the first trace frame.
        $throwSite = [
            'filename' => $error->getFile(),
            'lineno' => $error->getLine(),
            'in_app' => !str_contains($error->getFile(), '/vendor/'),
        ];

        if (isset($trace[0]['function'])) {
            $throwSite['function'] = $this->formatFunction($trace[0]);
        }

        $frames[] = $throwSite;

        return $frames;
    }

    /**
     * @param array{function: string, class?: class-string, type?: string} $frame
     */
    private function formatFunction(array $frame): string
    {
        return isset($frame['class'])
            ? $frame['class'] . ($frame['type'] ?? '::') . $frame['function']
            : $frame['function'];
    }

    private const array HANDLED_HTTP_KEYS = [
        'http.url',
        'http.method',
        'http.query',
        'http.response.status_code',
    ];

    /**
     * Classify attributes into tags, contexts, and extra based on the classifier.
     *
     * @param array<string, mixed> $attributes
     * @return array{array<string, string>, array<string, mixed>, array<string, mixed>}
     */
    private function classifyAttributes(array $attributes): array
    {
        $tags = [];
        $contexts = [];
        $extra = [];

        foreach ($attributes as $key => $value) {
            // Skip internal span attributes and level (handled in payload)
            if (str_starts_with((string) $key, 'span.')) {
                continue;
            }
            if ($key === 'level') {
                continue;
            }
            // Skip only the HTTP attributes we handle in buildRequest/buildResponse
            if (\in_array($key, self::HANDLED_HTTP_KEYS, true)) {
                continue;
            }

            $field = ($this->classifier)($key);

            switch ($field) {
                case SentryField::Tag:
                    // Tags must be strings, max 200 chars
                    if (\is_scalar($value) || $value === null) {
                        $tags[$key] = substr((string) $value, 0, 200);
                    }
                    break;
                case SentryField::Context:
                    $contexts[$key] = $value;
                    break;
                case SentryField::Extra:
                    $extra[$key] = $value;
                    break;
            }
        }

        return [$tags, $contexts, $extra];
    }

    /**
     * @param array<string, mixed> $attributes
     * @return array<string, mixed>
     */
    private function buildRequest(array $attributes): array
    {
        $request = [];

        if (isset($attributes['http.url']) && \is_string($attributes['http.url'])) {
            $request['url'] = $attributes['http.url'];
        }

        if (isset($attributes['http.method']) && \is_string($attributes['http.method'])) {
            $request['method'] = $attributes['http.method'];
        }

        if (isset($attributes['http.query']) && \is_string($attributes['http.query'])) {
            $request['query_string'] = $attributes['http.query'];
        }

        return $request;
    }

    /**
     * @param array<string, mixed> $attributes
     * @return array<string, mixed>
     */
    private function buildResponse(array $attributes): array
    {
        $response = [];

        if (isset($attributes['http.response.status_code']) && \is_int($attributes['http.response.status_code'])) {
            $response['status_code'] = $attributes['http.response.status_code'];
        }

        return $response;
    }
}
